### PR TITLE
[webgpu] Add depthwise and other 21 programs WGSL support

### PR DESCRIPTION
**Pull request by axinging**
Changed programs:
```
addn_packed_webgpu.ts
batchnorm_webgpu.ts
clip_vec4_webgpu.ts
clip_webgpu.ts
concat_webgpu.ts
crop_and_resize_webgpu.ts
depthwise_conv2d_webgpu.ts
fill_webgpu.ts
gather_nd_webgpu.ts
gather_webgpu.ts
mirror_pad_webgpu.ts
pad_webgpu.ts
pool2d_webgpu.ts
pool_filtersizeone_webgpu.ts
resize_bilinear_webgpu.ts
select_webgpu.ts
slice_webgpu.ts
strided_slice_webgpu.ts
tile_webgpu.ts
transform_webgpu.ts
transpose_shared_webgpu.ts
transpose_webgpu.ts
```

To test WGSL, change WEBGPU_USE_GLSL in tfjs-backend-webgpu/src/flags_webgpu.ts to true.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5403)

### CODE REVIEW

File: tfjs-backend-webgpu/src/kernels/select_webgpu.ts [hunk 1]
```diff
@@ -17,33 +17,34 @@
...
+import {getUseWgsl, WebGPUProgram} from './webgpu_program';
 
 export class SelectProgram implements WebGPUProgram {
   variableNames = ['c', 'a', 'b'];
   outputShape: number[];
   shaderKey: string;
   dispatchLayout: {x: number[]};
   dispatch: [number, number, number];
-  workPerThread = 4;
   workGroupSize: [number, number, number] = [16, 1, 1];
```
Comment (qjia7) [inline tfjs-backend-webgpu/src/kernels/select_webgpu.ts:31]: Please change the work group size to [64, 1, 1] instead of [16, 1, 1]

**Review by qjia7 — APPROVED** [review]
LGTM with nits.
@kainino0x @lina128 @jinjingforever take a look, thanks.

File: tfjs-backend-webgpu/src/kernels/clip_vec4_webgpu.ts [hunk 2]
```diff
@@ -59,4 +64,27 @@ export class ClipVec4Program implements WebGPUProgram {
...
+  getUserCodeWgsl(): string {
+    const userCode = `
+      ${getWorkGroupSizeStringWgsl(this.workGroupSize)}
+      fn main([[builtin(global_invocation_id)]] globalId : vec3<u32>) {
+        let index = globalId.x;
+        if(index < uniforms.size) {
+          let value = getAAtOutCoordsByGlobalId(globalId);
+          var clampedValue : vec4<f32>;
+          for (var i = 0u; i < 4u; i = i + 1u) {
+            if (isNanCustom(value[i])) {
```
[the remark this answers is not in the document]
Comment (kainino0x) [reply tfjs-backend-webgpu/src/kernels/clip_vec4_webgpu.ts:77]: btw when working with NaN, please be aware of this (I just happened to find this when checking the spec for isNan):

> - Implementations may assume that NaNs, inifinities are not present
>     Note: This means some functions (e.g. isInf, isNan, min and max) may not return the expected result due to optimizations about the presence of NaNs and infinities.



https://gpuweb.github.io/gpuweb/wgsl/#floating-point-evaluation

File: tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+import * as tf from '@tensorflow/tfjs-core';
+import {test_util} from '@tensorflow/tfjs-core';
+import {describeWebGPU} from '../test_util';
+
+const expectArraysClose = test_util.expectArraysClose;
+
+describeWebGPU('pool', () => {
+  // For PoolWithFilterSizeEqualsOneProgram. This case will fail wasm, so keep
+  // it here. Wasm bug: https://github.com/tensorflow/tfjs/issues/5471.
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts:26]: Is it possible to add this test generally and skip it on the wasm backend specifically?

**Review by kainino0x — APPROVED** [review]
As usual I can't review in detail but lgtm

File: tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts [hunk 4]
```diff
@@ -99,4 +103,46 @@ export class BatchNormProgram implements WebGPUProgram {
...
+      }
+      ${getWorkGroupSizeStringWgsl(this.workGroupSize)}
+      fn main([[builtin(global_invocation_id)]] globalId : vec3<u32>) {
+        let coords = getOutputCoords(globalId);
+        let xValue = getXAtOutCoordsByGlobalId(globalId);
+        let meanValue = getMeanAtOutCoordsByGlobalId(globalId);
+        let varianValue = getVarianceAtOutCoordsByGlobalId(globalId);
+        let offsetValue = ${offsetSnippet};
+        let scaleValue = ${scaleSnippet};
+        let inv = scaleValue * 1.0/sqrt(varianValue + f32(uniforms.varianceEpsilon));
```
[the remark this answers is not in the document]
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/batchnorm_webgpu.ts:142]: When I sumit my PR, not supported.

File: tfjs-backend-webgpu/src/kernels/depthwise_conv2d_webgpu.ts [hunk 5]
```diff
@@ -162,4 +169,113 @@ export class DepthwiseConv2DProgram implements WebGPUProgram {
...
+
+      fn writeResult(batch : u32, row : u32, col : u32, chan : u32, value : f32) {
+        let coord = vec4<u32>(batch, row, col, chan);
+        if (coordsInBounds4D(coord, uniforms.outShape)) {
+          setOutput(batch, row, col, chan, value);
+        }
+      }
+
+      ${getWorkGroupSizeStringWgsl(this.workGroupSize)}
+      fn main([[builtin(local_invocation_id)]] localId : vec3<u32>,
```
[the remark this answers is not in the document]
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/depthwise_conv2d_webgpu.ts:211]: Removed

File: tfjs-backend-webgpu/src/kernels/select_webgpu.ts (hunk 1 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/select_webgpu.ts:31]: done

File: tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts (hunk 3 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts:26]: I think this is possible. How about I follow up this in another PR?

Comment (kainino0x) [reply tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts:26]: OK, in that case perhaps leave a TODO here.

**Review by lina128 — APPROVED** [review]
LGTM from TFJS side.

> *[Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/5403#-:-MhQDNXq91NbAS7AxWkP:bpo5jz5)* status: :shipit: complete! 3 of 1 approvals obtained (waiting on @axinging, @jinjingforever, and @qjia7)

File: tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts (hunk 3 above)
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/pool_webgpu_test.ts:26]: Thanks, TODO added, and BUG is here: https://github.com/tensorflow/tfjs/issues/5506

File: tfjs-backend-webgpu/src/kernels/fill_webgpu.ts [hunk 6]
```diff
@@ -54,4 +60,20 @@ export class FillProgram implements WebGPUProgram {
...
+
+  getUserCodeWgsl(): string {
+    const userCode = `
+    ${getWorkGroupSizeStringWgsl(this.workGroupSize)}
+    fn main([[builtin(global_invocation_id)]] globalId : vec3<u32>) {
+      let index = globalId.x;
+      for (var i = 0u; i < ${this.workPerThread}u; i = i + 1u) {
+        let flatIndex = index * ${this.workPerThread}u + i;
+        if (flatIndex < uniforms.size) {
+          setOutputFlat(flatIndex, f32(uniforms.value));
```
[the remark this answers is not in the document]
Comment (axinging) [reply tfjs-backend-webgpu/src/kernels/fill_webgpu.ts]: Thansk, Removed.
